syft/messaging/promise.py

```diff
@@ -0,0 +1,165 @@
+"""This file contains the object we use to tell another worker that they will
+receive an object with a certain ID. This object will also contain a set of
+Plan objects which will be executed when the promise is kept (assuming
+the plan has all of the inputs it requires). """
+
+import syft as sy
+
+from abc import ABC
+from syft.workers.abstract import AbstractWorker
+
+
+class Promise(ABC):
+    def __init__(self, owner=None, id=None, obj_id=None, obj_type=None, plans=None):
+        """Initialize a Promise with a unique ID and a set of (possibly empty) plans
+
+        A Promise is a data-structure which indicates that "there will be an object
+        with this ID at some point, and when you get it, use it to execute these plans".
+
+        As such, the promise has an ID itself and it also has an queue of object ids
+        with which the promise has been kept.
+
+        However, it's important to know that some Plans are actually waiting on multiple
+        objects before they can be executed. Thus, it's possible that you might call
+        .keep() on a promise and nothing will happen because all of the plans are also
+        waiting on other promises to be kept before they execute.
+
+        Args:
+            id (int): the id of the promise
+            plans (set): a set of the plans waiting on the promise to be kept
+        Example:
+            future_x = Promise()
+        """
+
+        self.owner = owner
+
+        if id is None:
+            id = sy.ID_PROVIDER.pop()
+        self._id = id
+
+        self.obj_type = obj_type
+        self.queue_obj_ids = []
+
+        if plans is None:
+            plans = set()
+        self.plans = plans
+
+    def keep(self, obj):
+        """ This method is used to keep a promise.
+        This will register the object on the worker, add its id to the queue of the promise,
+        and every plan waiting for this promise will try to execute if it can.
+        """
+        if obj.type() != self.obj_type:
+            raise TypeError(
+                "keep() was called with an object of incorrect type (not the type that was promised)"
+            )
+
+        if self.id in self.owner._objects:
+            self.owner.register_obj(obj)
+
+        self.queue_obj_ids.append(obj.id)
+
+        # If some plans were waiting for this promise...
+        for plan_id in self.plans:
+            # If the promise has been moved, the plan waiting for it might not be on the same worker
+            # TODO is this ok?
+            if plan_id in self.owner._objects:
+                plan = self.owner._objects[plan_id]
```

Comment:
it's ok because the owner is support to be local, but use instead .get_obj()
`_objects` should only be used for debug purpose